ksql-execution/src/main/java/io/confluent/ksql/execution/plan/StreamSource.java

```diff
@@ -63,22 +99,359 @@ public boolean equals(final Object o) {
+        new DefaultExecutionStepProperties(
+            queryContext.toString(),
+            schemaWithMetaAndKeyFields.getSchema(),
+            queryContext
+        ),
+        topicName,
+        Formats.of(keyFormat, valueFormat, options),
+        timestampPolicy,
+        getTimeStampColumnIndex(timestampPolicy, schemaWithMetaAndKeyFields.getSchema()),
+        offsetReset,
+        schemaWithMetaAndKeyFields.getOriginalSchema(),
+        new Functions<>(
+            (fmt, schema, ctx, builder) -> builder.buildKeySerde(
+                fmt.getFormatInfo(),
+                schema,
+                ctx
+            ),
+            nonWindowedValueMapper(schemaWithMetaAndKeyFields.getSchema()),
+            new AddTimestampColumn(),
+            Consumed::with,
+            PhysicalSchema::from
+        )
+    );
+  }
+
+  public static StreamSource<Struct> createNonWindowed(
+      final QueryContext queryContext,
+      final LogicalSchemaWithMetaAndKeyFields schemaWithMetaAndKeyFields,
+      final String topicName,
+      final KeyFormat keyFormat,
+      final ValueFormat valueFormat,
+      final Set<SerdeOption> options,
+      final TimestampExtractionPolicy timestampPolicy,
+      final Optional<AutoOffsetReset> offsetReset) {
+    return createNonWindowed(
+        queryContext,
+        schemaWithMetaAndKeyFields,
+        topicName,
+        keyFormat,
+        valueFormat,
+        options,
+        timestampPolicy,
+        offsetReset,
+        StreamSource::new
+    );
+  }
+
+  private static org.apache.kafka.connect.data.Field getKeySchemaSingleField(
+      final LogicalSchema schema) {
+    if (schema.keySchema().fields().size() != 1) {
+      throw new IllegalStateException("Only single key fields are currently supported");
+    }
+    return schema.keySchema().fields().get(0);
+  }
+
+  private static ValueMapperWithKey<Windowed<Struct>, GenericRow, GenericRow> windowedMapper(
+      final LogicalSchema schema) {
+    final org.apache.kafka.connect.data.Field keyField = getKeySchemaSingleField(schema);
+    return (keyStruct, row) -> {
+      if (row != null) {
+        final Window window = keyStruct.window();
+        final long start = window.start();
+        final String end = window instanceof SessionWindow ? String.valueOf(window.end()) : "-";
+        final Object key = keyStruct.key().get(keyField);
+        final String rowKey = String.format("%s : Window{start=%d end=%s}", key, start, end);
+        row.getColumns().add(0, rowKey);
+      }
+      return row;
+    };
+  }
+
+  private static ValueMapperWithKey<Struct, GenericRow, GenericRow> nonWindowedValueMapper(
+      final LogicalSchema schema) {
+    final org.apache.kafka.connect.data.Field keyField = getKeySchemaSingleField(schema);
+    return (key, row) -> {
+      if (row != null) {
+        row.getColumns().add(0, key.get(keyField));
+      }
+      return row;
+    };
+  }
+
+  private static int getTimeStampColumnIndex(
```

Comment:
a little orthogonal to the point of this pr, but I feel like this makes more sense inside `LogicalSchema` than `StreamSource`. As it stands I'm a little confused as to what `StreamSource` is supposed to be responsible for

Reply:
> I feel like this makes more sense inside LogicalSchema than StreamSource

That actually makes a lot of sense, thanks!

> I'm a little confused as to what StreamSource is supposed to be responsible for

`StreamSource` is supposed to build the part of the streams app that loads a KStream from a topic, and sets up the row to match the schema of the data source (by adding meta fields to the value). We could maybe split it into 2 steps - I'm kind of indifferent.

Reply:
We can keep it as is for now and if it gets unwieldily we can split it - thanks for the clarification!